### PR TITLE
[Graphbolt] Add dgl minibatch

### PR DESCRIPTION
**Pull request by peizhou001**
## Description
Add dgl minibatch used for computing.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: python/dgl/graphbolt/dgl_minibatch.py
Comment (frozenbugs) [inline python/dgl/graphbolt/dgl_minibatch.py:1]: nit: I would just put it in python/dgl/graphbolt/minibatch.py.

File: python/dgl/graphbolt/dgl_minibatch.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    necessary fields for computation using the DGL library.."""
+
+    blocks: List[DGLBlock] = None
+    """A list of 'DGLBlock's, each one corresponding to one layer, representing
+    a bipartite graph used for message passing.
+    """
+
+    input_nodes: Union[torch.Tensor, Dict[str, torch.Tensor]] = None
+    """A representation of input nodes in the outermost layer. Conatins all nodes
+       in the 'sampled_subgraphs'.
```
Comment (frozenbugs) [inline python/dgl/graphbolt/dgl_minibatch.py:26]:
```suggestion
    """A representation of input nodes in the outermost layer. Contains all
       nodes in the 'sampled_subgraphs'.
```

File: python/dgl/graphbolt/dgl_minibatch.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    - If `input_nodes` is a tensor: It indicates the graph is homogeneous.
+    - If `input_nodes` is a dictionary: The keys should be node type and the
+      value should be corresponding heterogeneous node id.
+    """
+
+    output_nodes: Union[torch.Tensor, Dict[str, torch.Tensor]] = None
+    """
+    Representation of output nodes, usually also the seed nodes, used for
```
Comment (frozenbugs) [inline python/dgl/graphbolt/dgl_minibatch.py:34]: Make it on the same row? consistent with the others.

File: python/dgl/graphbolt/dgl_minibatch.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+      - If keys are single strings: It means the graph is homogeneous, and the
+      keys are feature names.
+      - If keys are tuples: It means the graph is heterogeneous, and the keys
+      are tuples of '(node_type, feature_name)'.
+    """
+
+    edge_features: List[
+        Union[Dict[str, torch.Tensor], Dict[Tuple[str, str], torch.Tensor]]
+    ] = None
+    """Edge features associated with the 'sampled_subgraphs'.
```
Comment (frozenbugs) [inline python/dgl/graphbolt/dgl_minibatch.py:54]: associated with the 'blocks'.

**Review by frozenbugs — APPROVED** [review]

File: python/dgl/graphbolt/minibatch.py [hunk 4]
```diff
@@ -6,11 +6,97 @@
...
 
 from .base import etype_str_to_tuple
 from .sampled_subgraph import SampledSubgraph
 
-__all__ = ["MiniBatch"]
+__all__ = ["DGLMiniBatch", "MiniBatch"]
+
+
+@dataclass
+class DGLMiniBatch:
```
Comment (frozenbugs) [inline python/dgl/graphbolt/minibatch.py:18]: Move this blow Minibatch.

File: python/dgl/graphbolt/minibatch.py [hunk 5]
```diff
@@ -6,11 +6,97 @@
...
+      - If keys are single strings: It means the graph is homogeneous, and the
+      keys are feature names.
+      - If keys are tuples: It means the graph is heterogeneous, and the keys
+      are tuples of '(edge_type, feature_name)'. Note, edge type is a triplet
+      of format (str, str, str).
+    """
+
+    labels: Union[torch.Tensor, Dict[str, torch.Tensor]] = None
+    """
+    Labels associated with seed nodes / node pairs in the graph.
```
Comment (frozenbugs) [inline python/dgl/graphbolt/minibatch.py:66]:
```suggestion
    """Labels associated with seed nodes / node pairs in the graph.
```

File: python/dgl/graphbolt/minibatch.py [hunk 6]
```diff
@@ -6,11 +6,97 @@
...
+      should be corresponding labels to given 'output_nodes' or 'node_pairs'.
+    """
+
+    positive_node_pairs: Union[
+        Tuple[torch.Tensor, torch.Tensor],
+        Dict[str, Tuple[torch.Tensor, torch.Tensor]],
+    ] = None
+    """
+    Representation of positive graphs used for evaluating or computing loss in
+    link prediction tasks.
```
Comment (frozenbugs) [inline python/dgl/graphbolt/minibatch.py:79]:
```suggestion
    """Representation of positive graphs used for evaluating or computing loss
    in link prediction tasks.
```

File: python/dgl/graphbolt/minibatch.py [hunk 7]
```diff
@@ -6,11 +6,97 @@
...
+    given type.
+    """
+
+    negative_node_pairs: Union[
+        Tuple[torch.Tensor, torch.Tensor],
+        Dict[str, Tuple[torch.Tensor, torch.Tensor]],
+    ] = None
+    """
+    Representation of negative graphs used for evaluating or computing loss in
+    link prediction tasks.
```
Comment (frozenbugs) [inline python/dgl/graphbolt/minibatch.py:93]:
```suggestion
    """Representation of negative graphs used for evaluating or computing loss
    in link prediction tasks.
```

**Review by frozenbugs — APPROVED** [review]

File: python/dgl/graphbolt/minibatch.py (hunk 4 above)
Comment (peizhou001) [reply python/dgl/graphbolt/minibatch.py:18]: MIniBatch need to use it so it has to be defined ahead.

**Pull request merged**